### PR TITLE
shell: Fix memory leak, lazily create `ShellInterpreter` object

### LINKED ISSUE #11816: Memory Leak in `Bun.$` (ShellInterpreter)
### What version of Bun is running?

1.1.12+43f0913c3

### What platform is your computer?

Linux 6.8.4-3-pve x86_64

### What steps can reproduce the bug?

```ts
import { gc, $ } from "bun";
import { heapStats } from "bun:jsc";
for (let j = 0; j < 10; j++) {
  const promises = [];
  for (let i = 0; i < 10; i++) {
    promises.push(
      $`/opt/homebrew/opt/coreutils/libexec/gnubin/cat /tmp/over/input.txt`
    );
  }

  await Promise.all(promises);
  gc(true);
}

console.error("ShellInterpreter count", heapStats().objectTypeCounts.ShellInterpreter);
console.error("RSS", process.memoryUsage.rss() / 1024 / 1024 | 0, "MB");
```
where `/tmp/over/input` is `10 MB` from `/dev/urandom`.
kindly provided by @Jarred-Sumner 

### What is the expected behavior?

Finished ShellInterpreter should be garbage collected without any side effects.

### What do you see instead?

When the ShellInterpreter finishes the execution it retains the memory and does not release it. Even explicit call to the `Bun.gc` does not collect the retained memory. Depending on frequency of use it can grow up to `+1 GB` per 15-20 mins.

### Additional information

I have Callgrind and DHAT data available for a bigger project for indepth analysis. Please request fom me if it's necessary.

### PR DESCRIPTION
**Pull request by zackradisic**
### What does this PR do?

Fixes #11816

**Problem**
- ShellPromise's `resolve/reject` functions closes over `ShellInterpreter` class
- the `ShellInterpreter` class holds strong references to `resolve/reject` functions
- this makes it possible for the `ShellInterpreter` class to never be deinitialized
- was not caught in shell leak tests because of immediately immediately invoked function syntax: `(async function() { /* shell stuff */ })()`

**Fix**
- lazily create `ShellInterpreter` when the `ShellPromise` is actually awaited
- instead of giving `resolve/reject` to `ShellInterpreter`, give it to new `ParsedShellScript` class which holds script AST and other options
- this breaks cycle: now `resolve/reject` don't refer to `ShellInterpreter`
- put `values: ["resolve", "reject"]` in the `ShellInterpreter` class generator
- give `ShellInterpreter` `resolve/reject` when it gets created

### CODE REVIEW

File: src/shell/interpreter.zig [hunk 1]
```diff
@@ -602,11 +602,27 @@ pub const EnvMap = struct {
...
+pub const ShellArgs = struct {
+    /// This is the arena used to allocate the input shell script's AST nodes,
+    /// tokens, and a string pool used to store all strings.
+    arena: bun.ArenaAllocator,
+    /// Root ast node
+    script_ast: ast.Script = .{ .stmts = &[_]ast.Stmt{} },
+
+    /// create new ShellArgs
+    /// the ShellArgs struct itself is allocated in the arena
+    pub fn init(allocator: Allocator) *ShellArgs {
```
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig]:
```zig
arena: *bun.ArenaAllocator,

pub usingnamespace bun.New(@This());

pub fn allocator(this: *ShellArgs) std.mem.Allocator {
  return this.arena.allocator();
}

pub fn deinit(this: *ShellArgs) void {
  this.arena.deinit();
  this.destroy();
}

pub fn init() *ShellArgs {
  const arena = bun.new(bun.ArenaAllocator);
  arena.* = bun.ArenaAllocator.init(bun.default_allocator);
  return ShellArgs.new(.{
    .arena = arena,
   });
});
```

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
Test failures

**Comment by Zamion101** [comment]
Looks like some tests are failing but they're not associated with this PR. Is it possible to get a Canary for this PR or must all the fails be adressed before it's possible.

**Comment by Jarred-Sumner** [comment]
@zackradisic bunshell tests are failing on Windows. I think that is a blocker here.

File: src/shell/interpreter.zig [hunk 2]
```diff
@@ -602,19 +602,233 @@ pub const EnvMap = struct {
...
+
+    pub usingnamespace bun.New(@This());
+
+    pub fn arena_allocator(this: *ShellArgs) std.mem.Allocator {
+        return this.__arena.allocator();
+    }
+
+    pub fn deinit(this: *ShellArgs) void {
+        this.__arena.deinit();
+        this.destroy();
```
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig:621]: This leaks the bun.ArenaAllocator pointer.


```suggestion
        bun.destroy(this.__arena);
        this.destroy();
```

File: src/shell/interpreter.zig [hunk 3]
```diff
@@ -602,19 +602,234 @@ pub const EnvMap = struct {
...
+    pub fn createParsedShellScript(
+        globalThis: *JSC.JSGlobalObject,
+        callframe: *JSC.CallFrame,
+    ) callconv(.C) JSValue {
+        var shargs = ShellArgs.init();
+
+        const arguments_ = callframe.arguments(2);
+        var arguments = JSC.Node.ArgumentsSlice.init(globalThis.bunVM(), arguments_.slice());
+        const string_args = arguments.nextEat() orelse {
+            globalThis.throw("shell: expected 2 arguments, got 0", .{});
```
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig]: Memory leak:

```suggestion
            globalThis.throw("shell: expected 2 arguments, got 0", .{});
            shargs.deinit();
```

File: src/shell/interpreter.zig [hunk 4]
```diff
@@ -602,19 +602,234 @@ pub const EnvMap = struct {
...
+    }
+
+    pub fn createParsedShellScript(
+        globalThis: *JSC.JSGlobalObject,
+        callframe: *JSC.CallFrame,
+    ) callconv(.C) JSValue {
+        var shargs = ShellArgs.init();
+
+        const arguments_ = callframe.arguments(2);
+        var arguments = JSC.Node.ArgumentsSlice.init(globalThis.bunVM(), arguments_.slice());
```
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig]: This leaks memory on errors.

Let's just not use ArgumentsSlice here - it's unnecessary. And, let's create the shell args after validating the arguments. 

```suggestion
        const arguments = arguments_.slice();
        
       if (arguments.len < 2) {
            globalThis.throwNotEnoughArguments("Bun.$", 2, arguments.len);
            return .zero;
        }
        const string_args = arguments[0];
        const template_args_jst = arguments[1];
```

File: src/shell/interpreter.zig [hunk 5]
```diff
@@ -936,100 +1158,87 @@ pub const Interpreter = struct {
...
         };
 
-        script_heap.* = script_ast;
+        var shargs: *ShellArgs = undefined;
+        var jsobjs: std.ArrayList(JSValue) = std.ArrayList(JSValue).init(allocator);
+        var quiet: bool = false;
+        var cwd: ?bun.String = null;
+        var export_env: ?EnvMap = null;
+
+        parsed_shell_script.take(
```
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig:1198]: What verifies that the shell args in `parsed_shell_scripts` is not null? If it was null, wouldn't this be undefined memory and cause a segfault?

File: src/bun.js/api/ShellArgs.classes.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+export default [
+  define({
+    name: "ParsedShellScript",
+    construct: true,
+    noConstructor: true,
+    finalize: true,
+    hasPendingActivity: false,
+    configurable: false,
+    klass: {},
+    JSType: "0b11101110",
```
Comment (Jarred-Sumner) [inline src/bun.js/api/ShellArgs.classes.ts]:
```suggestion

```

File: src/bun.js/api/Shell.classes.ts [hunk 7]
```diff
@@ -4,11 +4,13 @@ export default [
   define({
     name: "ShellInterpreter",
     construct: true,
+    noConstructor: true,
     finalize: true,
     hasPendingActivity: true,
     configurable: false,
     klass: {},
     JSType: "0b11101110",
```
Comment (Jarred-Sumner) [inline src/bun.js/api/Shell.classes.ts]:
```suggestion

```

**Pull request merged**